packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPEventBridge.m

```diff
@@ -0,0 +1,122 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "./include/video_player_avfoundation/FVPEventBridge.h"
+
+#import <Foundation/Foundation.h>
+
+#if TARGET_OS_OSX
+#import <FlutterMacOS/FlutterMacOS.h>
+#else
+#import <Flutter/Flutter.h>
+#endif
+
+@interface FVPEventBridge () <FlutterStreamHandler>
+
+/// The event channel to dispatch notifications to.
+// TODO(stuartmorgan): Convert this to Pigeon event channels once the plugin is using Swift
+// Pigeon generation.
+@property(nonatomic) FlutterEventChannel *eventChannel;
+
+/// The event sink associated with eventChannel.
+///
+/// Will be nil both before the channel listener is ready on the Dart side, and after it has been
+/// closed on the Dart side.
+@property(nonatomic, nullable) FlutterEventSink eventSink;
+
+/// A queue of events received before eventSink is ready, to dispatch once the channel is fully
+/// set up.
+@property(nonatomic) NSMutableArray<NSObject *> *queuedEvents;
+
+@end
+
+@implementation FVPEventBridge
+
+- (instancetype)initWithMessenger:(NSObject<FlutterBinaryMessenger> *)messenger
+                      channelName:(NSString *)channelName {
+  self = [super init];
+  if (self) {
+    _queuedEvents = [[NSMutableArray alloc] init];
+    _eventChannel = [FlutterEventChannel eventChannelWithName:channelName
+                                              binaryMessenger:messenger];
+    // This retain loop is broken in videoPlayerWasDisposed.
+    [_eventChannel setStreamHandler:self];
+  }
+  return self;
+}
+
+#pragma mark FlutterStreamHandler
+
+- (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments
+                                       eventSink:(nonnull FlutterEventSink)events {
+  self.eventSink = events;
+
+  // Send any events that came in before the sink was ready.
+  for (id event in self.queuedEvents) {
+    self.eventSink(event);
+  }
+  [self.queuedEvents removeAllObjects];
+
+  return nil;
+}
+
+- (FlutterError *_Nullable)onCancelWithArguments:(id _Nullable)arguments {
```

Comment:
The documentation says:

> Such request pairs may occur during Flutter hot restart.

Are plugins destroyed and recreated during hot restart, or they live through hot restart? If it's the latter then the event queue is always nil after a hot restart?

Reply:
The native code's plugin instance survives hot restart, but `video_player` calls an init method when the Dart plugin instance (which doesn't survive) is created, which will call [this code](https://github.com/flutter/packages/blob/7a9e9ceee821c13e851a09c65d3e9da38564748f/packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m#L184-L185) and destroy all the player instances, and thus their bridges.

(Even if it didn't, since these channels have ID-based names, and the IDs are assigned by the native code, after a hot restart it should never be possible for the Dart code to try re-listening to any previously-existing video player event channels anyway.)